### PR TITLE
[AST/Sema] NonCopyableGenerics: Address some of the request circularity issues

### PR DESCRIPTION
**Pull request by xedin**
- Extract logic related to protocol handling out of `InvertibleAnnotationRequest`
- Don't resolve requirements that don't have inverse as a constraint

### CODE REVIEW

File: include/swift/AST/Decl.h [hunk 1]
```diff
@@ -5216,6 +5216,11 @@ class ProtocolDecl final : public NominalTypeDecl {
   /// protocol.
   bool inheritsFrom(const ProtocolDecl *Super) const;
 
+  /// Determine whether this protocol has ~<target>` stated on
+  /// itself, one of its inherited types or `Self` requirements.
+  std::pair</*found=*/bool, /*where=*/SourceLoc>
+  hasInverseMarking(InvertibleProtocolKind ip) const;
```
Comment (kavon) [inline include/swift/AST/Decl.h:5222]: if CI fails, this would be nice to fix: name parameter `target` instead of `ip`.

**Review by kavon — APPROVED** [review]

File: include/swift/AST/Decl.h (hunk 1 above)
Comment (xedin) [reply include/swift/AST/Decl.h]: Done!

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
Please test with following pull request:
https://github.com/apple/swift-docc/pull/836

@swift-ci please test Linux platform

**Comment by xedin** [comment]
@swift-ci please test Linux

**Comment by xedin** [comment]
@swift-ci please test macOS platform

**Comment by xedin** [comment]
@swift-ci please clean test macOS platform

**Comment by xedin** [comment]
@swift-ci please smoke test macOS platform

**Pull request merged**